File: CS/WebFormsDashboardCustomPropertiesSample/SessionDashboardStorage.cs
using System;
using System.Collections.Generic;
using System.Xml.Linq;
using DevExpress.DashboardCommon;
using DevExpress.DashboardWeb;
using System.Web;
using System.Web.SessionState;

public class SessionDashboardStorage : DashboardStorageBase {
    const string DashboardStorageKey = "DashboardStorage";

    private static SessionDashboardStorage instance = null;

    public static SessionDashboardStorage Instance {
        get {
            if(instance == null) {
                instance = new SessionDashboardStorage();
            }
            return instance;
        }
    }

    Dictionary<string, XDocument> Storage {
        get {
            HttpSessionState session = HttpContext.Current.Session;
            if(session != null) {
                Dictionary<string, XDocument> storage = session[DashboardStorageKey] as Dictionary<string, XDocument>;
                if(storage == null) {
                    storage = new Dictionary<string, XDocument>();
                    session[DashboardStorageKey] = storage;
                    return storage;
                }
                return storage;
            }
            throw new Exception();
        }
    }

    protected SessionDashboardStorage() : base() {

    }

    protected override IEnumerable<string> GetAvailableDashboardsID() {
        return Storage.Keys;
    }
    protected override XDocument LoadDashboard(string dashboardID) {
        XDocument document = Storage[dashboardID];
        HttpContext httpContext = HttpContext.Current;
        if(dashboardID == "ProductDetails" && httpContext != null) {
            Dashboard dashboard = new Dashboard();
            dashboard.LoadFromXDocument(document);
            string applicationPath = httpContext.Request.ApplicationPath.TrimEnd('/') + "/";
            BoundImageDashboardItem primaryImage = dashboard.Items["primaryImage"] as BoundImageDashboardItem;
            if(primaryImage != null)
                primaryImage.UriPattern = applicationPath + "Content/ProductDetailsImages/{0}.jpg";
            BoundImageDashboardItem secondaryImage = dashboard.Items["secondaryImage"] as BoundImageDashboardItem;
            if(secondaryImage != null)
                secondaryImage.UriPattern = applicationPath + "Content/ProductDetailsImages/{0} Secondary.jpg";
            document = dashboard.SaveToXDocument();
        }
        return document;
    }
    protected override void SaveDashboard(string dashboardID, XDocument dashboard, bool createNew) {
        if(createNew ^ Storage.ContainsKey(dashboardID))
            Storage[dashboardID] = dashboard;
    }
    public void RegisterDashboard(string dashboardID, XDocument dashboard) {
        SaveDashboard(dashboardID, dashboard, true);
    }
}

File: VB/WebFormsDashboardCustomPropertiesSample/SessionDashboardStorage.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Xml.Linq
Imports DevExpress.DashboardCommon
Imports DevExpress.DashboardWeb
Imports System.Web
Imports System.Web.SessionState

Public Class SessionDashboardStorage
	Inherits DashboardStorageBase
	Private Const DashboardStorageKey As String = "DashboardStorage"

	Private Shared instance_Renamed As SessionDashboardStorage = Nothing

	Public Shared ReadOnly Property Instance() As SessionDashboardStorage
		Get
			If instance_Renamed Is Nothing Then
				instance_Renamed = New SessionDashboardStorage()
			End If
			Return instance_Renamed
		End Get
	End Property

	Private ReadOnly Property Storage() As Dictionary(Of String, XDocument)
		Get
			Dim session As HttpSessionState = HttpContext.Current.Session
			If session IsNot Nothing Then
'INSTANT VB NOTE: The local variable storage was renamed since Visual Basic will not allow local variables with the same name as their enclosing function or property:
				Dim storage_Renamed As Dictionary(Of String, XDocument) = TryCast(session(DashboardStorageKey), Dictionary(Of String, XDocument))
				If storage_Renamed Is Nothing Then
					storage_Renamed = New Dictionary(Of String, XDocument)()
					session(DashboardStorageKey) = storage_Renamed
					Return storage_Renamed
				End If
				Return storage_Renamed
			End If
			Throw New Exception()
		End Get
	End Property

	Protected Sub New()
		MyBase.New()

	End Sub

	Protected Overrides Function GetAvailableDashboardsID() As IEnumerable(Of String)
		Return Storage.Keys
	End Function
	Protected Overrides Function LoadDashboard(ByVal dashboardID As String) As XDocument
		Dim document As XDocument = Storage(dashboardID)
		Dim httpContext As HttpContext = HttpContext.Current
		If dashboardID = "ProductDetails" AndAlso httpContext IsNot Nothing Then
			Dim dashboard As New Dashboard()
			dashboard.LoadFromXDocument(document)
			Dim applicationPath As String = httpContext.Request.ApplicationPath.TrimEnd("/"c) & "/"
			Dim primaryImage As BoundImageDashboardItem = TryCast(dashboard.Items("primaryImage"), BoundImageDashboardItem)
			If primaryImage IsNot Nothing Then
				primaryImage.UriPattern = applicationPath & "Content/ProductDetailsImages/{0}.jpg"
			End If
			Dim secondaryImage As BoundImageDashboardItem = TryCast(dashboard.Items("secondaryImage"), BoundImageDashboardItem)
			If secondaryImage IsNot Nothing Then
				secondaryImage.UriPattern = applicationPath & "Content/ProductDetailsImages/{0} Secondary.jpg"
			End If
			document = dashboard.SaveToXDocument()
		End If
		Return document
	End Function
	Protected Overrides Sub SaveDashboard(ByVal dashboardID As String, ByVal dashboard As XDocument, ByVal createNew As Boolean)
		If createNew Xor Storage.ContainsKey(dashboardID) Then
			Storage(dashboardID) = dashboard
		End If
	End Sub
	Public Sub RegisterDashboard(ByVal dashboardID As String, ByVal dashboard As XDocument)
		SaveDashboard(dashboardID, dashboard, True)
	End Sub
End Class